autonomy_core/map_plan/mapper/src/test_voxel_mapper.cpp

```diff
@@ -0,0 +1,428 @@
+#include <iostream>
+#include <gtest/gtest.h>
+#include <memory>
+#include <cmath>
+#include <chrono>
+#include <planning_ros_msgs/VoxelMap.h>
+#include "mapper/voxel_mapper.h"
+
+
+// Test class needs to be in the same namespace as VoxelMapper to be able to make it a friend class
+namespace mapper
+{
+
+/*** Declaration of fixture class to use the same configuration across tests ***/
+class VoxelMapperTest : public testing::Test
+{
+protected:
+    // Google test needs to create objects of this test fixture and it calls the default constructor
+    VoxelMapperTest();
+    VoxelMapperTest(double x_orig, double y_orig, double z_orig, int x_dim, int y_dim, int z_dim,
+                    double res, int decay_times);
+
+    const double x_origin_; 
+    const double y_origin_; 
+    const double z_origin_; 
+    const int x_dim_;
+    const int y_dim_;
+    const int z_dim_;
+    const double resolution_;
+    const int decay_times_;
+
+    // This is the VoxelMapper object that is used for all the tests
+    std::unique_ptr<mapper::VoxelMapper> p_test_mapper_;
```

Comment:
Maybe add a default ctor to VoxelMapper to avoid this unique_ptr stuff.

Reply:
Initially I didn't want to use the unique_ptr, but I was getting a "free(): invalid pointer" error and I couldn't figure out why; I was able to get around it and get things working with the unique_ptr. 

This time around I decided to give it another shot without the unique_ptr and I figured out that the issue lied in reassigning the mapper::VoxelMapper object. Since it was previously initialized with the default constructor, I would then reinitialize it in the SetUp() funcion like `test_mapper_ = mapper::Voxel(/*arguments here*/)` which made a shallow copy. I came up with two solutions

Method 1: Overload the assignment operator so that there is a proper copy
```
class VoxelMapperTest : public testing::Test
{
    // No intialization, just declaration
    mapper::VoxelMapper test_mapper_;
};

VoxelMapperTest::SetUp()
{
    test_mapper_ = mapper::Voxel(/*arguments here*/)
}

// In Voxel Mapper Class source file
void VoxelMapper::operator=(const VoxelMapper &vox_map)
{
    // Do correct copying here
}
```

Method 2: I could also not use the SetUp() function at all and just initialize everthing in the class declaration
```
class VoxelMapperTest : public testing::Test
{
    // Initialize other variables like
    double x_origin_ = -100;
    // also initialize the VoxelMapper object here
    mapper::VoxelMapper test_mapper_ =
        mapper::VoxelMapper(/*previosly initialized arguments from above*/);
};
```
Which method should be preferred?

Reply:
What is being shallow copied?

Reply:
I looked more into, the issue is the copying but it looks like its when copying the boost array. According to the documentation on the assignment operator for the boost multiarray [reference](https://www.boost.org/doc/libs/1_61_0/libs/multi_array/doc/reference.html#multi_array_ref:~:text=Modifiers.%EF%BF%BD-,multi_array%26%20operator%3D(const%20multi_array%26%20x)%3B,-template%20%3Cclass%20Array), one precondition is that all the dimensions need to be the same. If they aren't then the map isn't copied properley. I confirmed this by verifying the shape of the array before and after assignment; when I have the overloaded assignemnt operator I adjust the dimensions and I no longer get the error.

Reply:
That’s why I suggested using a simple vector

Reply:
but for, just to wrap up this PR. what would be better to avoid this "free() invalid pointer" error? I could leave the unique_ptr, add the overloading of the assignment operator for the voxel class or just intialize it in the test class declaration and not reinitialize it to avoid the copying issue

Reply:
Just use the unique_ptr for now.

Reply:
Ok, should I also delete the overloading and the default constructor that I added or should I leave it in?